src/handleRequest.ts

```diff
@@ -14,79 +14,97 @@ import {
+      const parsedUrl = new URL(req.url)
+      req.query = parsedUrl.searchParams
 
-    // Retrieve request URL parameters based on the provided mask
-    const params =
-      (mask && match(resolveRelativeUrl(mask), req.url).params) || {}
+      const relevantRequestHandler = requestHandlers.find((requestHandler) => {
+        return requestHandler.predicate(req, parsedUrl)
+      })
 
-    const requestWithParams: MockedRequest = {
-      ...req,
-      params,
-    }
+      if (relevantRequestHandler == null) {
+        return channel.send({ type: 'MOCK_NOT_FOUND' })
+      }
 
-    const context = defineContext
-      ? defineContext(requestWithParams)
-      : defaultContext
+      const { mask, defineContext, resolver } = relevantRequestHandler
 
-    const mockedResponse: MockedResponse | undefined = await resolver(
-      requestWithParams,
-      response,
-      context,
-    )
+      // Retrieve request URL parameters based on the provided mask
+      const params =
+        (mask && match(resolveRelativeUrl(mask), req.url).params) || {}
 
-    if (!mockedResponse) {
-      console.warn(
-        '[MSW] Expected a mocking resolver function to return a mocked response Object, but got: %s. Original response is going to be used instead.',
-        mockedResponse,
-      )
+      const requestWithParams: MockedRequest = {
+        ...req,
+        params,
+      }
 
-      return channel.send('MOCK_NOT_FOUND')
-    }
+      const context = defineContext
+        ? defineContext(requestWithParams)
+        : defaultContext
 
-    // Transform Headers into a list to be stringified preserving multiple
-    // header keys. Stringified list is then parsed inside the ServiceWorker.
-    const responseWithHeaders = {
-      ...mockedResponse,
-      // @ts-ignore
-      headers: Array.from(mockedResponse.headers.entries()),
-    }
+      const mockedResponse: MockedResponse | undefined = await resolver(
+        requestWithParams,
+        response,
+        context,
+      )
 
-    channel.send(JSON.stringify(responseWithHeaders))
+      if (!mockedResponse) {
+        console.warn(
+          '[MSW] Expected a mocking resolver function to return a mocked response Object, but got: %s. Original response is going to be used instead.',
+          mockedResponse,
+        )
+
+        return channel.send({ type: 'MOCK_NOT_FOUND' })
+      }
+
+      // Transform Headers into a list to be stringified preserving multiple
+      // header keys. Stringified list is then parsed inside the ServiceWorker.
+      const responseWithHeaders = {
+        ...mockedResponse,
+        // @ts-ignore
+        headers: Array.from(mockedResponse.headers.entries()),
+      }
+
+      channel.send({
+        type: 'MOCK_SUCCESS',
+        payload: responseWithHeaders,
+      })
+    } catch (error) {
+      channel.send({
+        type: 'INTERNAL_ERROR',
+        payload: {
+          status: 500,
```

Comment:
Could this be: `error.status || 500`?

Reply:
Hm. Could you please give me an example of such error?

Please note that throwing in request handlers is not the recommended way of mocking server errors. I'd strongly recommend mocking server errors according to [this guide](https://redd.gitbook.io/msw/recipes/mocking-error-responses), which implies composing an error response via `res()`.